subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/internal/precompiled/PrecompileGroovyScriptsTask.java

```diff
@@ -0,0 +1,299 @@
+                                       List<PrecompiledGroovyScript> scriptPlugins,
+                                       FileCollection classpath,
+                                       Provider<Directory> javaSourceDependencyClasses,
+                                       Provider<Directory> groovySourceDependencyClasses) {
+        this.scriptCompilationHandler = scriptCompilationHandler;
+        this.compileOperationFactory = compileOperationFactory;
+        this.fileSystemOperations = fileSystemOperations;
+        this.scriptRunnerFactory = scriptRunnerFactory;
+        this.serviceRegistry = serviceRegistry;
+
+        this.classLoaderScope = classLoaderScopeRegistry.getCoreAndPluginsScope();
+
+        this.scriptPlugins = scriptPlugins;
+        this.classpath = classpath;
+
+        DirectoryProperty buildDir = project.getLayout().getBuildDirectory();
+        this.intermediatePluginBlockClassesDir = buildDir.dir("groovy-dsl-plugins/plugin-blocks");
+        this.intermediatePluginClassesDir = buildDir.dir("groovy-dsl-plugins/work/classes");
+        this.intermediatePluginMetadataDir = buildDir.dir("groovy-dsl-plugins/work/metadata");
+
+        this.precompiledGroovyScriptsOutputDir.set(buildDir.dir("groovy-dsl-plugins/output/plugin-classes"));
+        this.pluginAdapterSourcesOutputDir.set(buildDir.dir("groovy-dsl-plugins/output/adapter-src"));
+        this.pluginAdapterClassesOutputDir.set(buildDir.dir("groovy-dsl-plugins/output/adapter-classes"));
+
+        this.javaSourceDependencyClasses = javaSourceDependencyClasses;
+        this.groovySourceDependencyClasses = groovySourceDependencyClasses;
+    }
+
+    @PathSensitive(PathSensitivity.RELATIVE)
+    @InputFiles
+    Set<File> getScriptFiles() {
+        return scriptPlugins.stream().map(p -> p.getSource().getResource().getFile()).collect(Collectors.toSet());
+    }
+
+    @PathSensitive(PathSensitivity.RELATIVE)
+    @InputFiles
+    Provider<Directory> getJavaSourceDependencyClasses() {
+        return javaSourceDependencyClasses;
+    }
+
+    @PathSensitive(PathSensitivity.RELATIVE)
+    @InputFiles
+    Provider<Directory> getGroovySourceDependencyClasses() {
+        return groovySourceDependencyClasses;
+    }
+
+    @Classpath
+    public FileCollection getClasspath() {
+        return classpath;
+    }
+
+    @OutputDirectory
+    Provider<File> getPrecompiledGroovyScriptsOutputDir() {
+        return precompiledGroovyScriptsOutputDir.getAsFile();
+    }
+
+    @OutputDirectory
+    DirectoryProperty getPluginAdapterSourcesOutputDir() {
+        return pluginAdapterSourcesOutputDir;
+    }
+
+    @Internal
+    Provider<File> getAdapterClassesOutputDir() {
+        return pluginAdapterClassesOutputDir.getAsFile();
+    }
+
+    @TaskAction
+    void compileScripts() {
+        FileCollection compileClasspath = classpath.plus(project.files(javaSourceDependencyClasses, groovySourceDependencyClasses));
```

Comment:
I think instead of wiring these in through separate properties, the compileClasspath should be a single property that's set in the plugin.  This makes it easier to completely swap out what the classpath should be without relying on built-in assumptions in this task.